homeassistant/components/gree/bridge.py

```diff
@@ -1,11 +1,197 @@
 """Helper and wrapper classes for Gree module."""
+from datetime import timedelta
+import logging
 from typing import List
 
 from greeclimate.device import Device, DeviceInfo
 from greeclimate.discovery import Discovery
-from greeclimate.exceptions import DeviceNotBoundError
+from greeclimate.exceptions import DeviceNotBoundError, DeviceTimeoutError
 
 from homeassistant import exceptions
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .const import DOMAIN, MAX_ERRORS
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class DeviceDataUpdateCoordinator(DataUpdateCoordinator):
+    """Manages polling for state changes from the device."""
+
+    def __init__(self, hass: HomeAssistant, device: Device):
+        """Initialize the data update coordinator."""
+        DataUpdateCoordinator.__init__(
+            self,
+            hass,
+            _LOGGER,
+            name=f"{DOMAIN}-{device.device_info.name}",
+            update_interval=timedelta(seconds=60),
+        )
+        self._device = device
+        self._error_count = 0
+
+    async def _async_update_data(self):
+        """Update the state of the device."""
+        try:
+            await self._device.update_state()
+
+            if not self.last_update_success and self._error_count:
+                _LOGGER.warning(
+                    "Device is available: %s (%s)",
+                    self.name,
+                    str(self._device.device_info),
+                )
+
+            self._error_count = 0
+        except DeviceTimeoutError as error:
+            self._error_count += 1
+
+            # Under normal conditions GREE units timeout every once in a while
+            if self.last_update_success and self._error_count >= MAX_ERRORS:
+                _LOGGER.warning(
+                    "Device is unavailable: %s (%s)",
+                    self.name,
+                    self._device.device_info,
+                )
+                raise UpdateFailed(error) from error
+
+    async def push_state_update(self):
+        """Send state updates to the physical device."""
+        try:
+            return await self._device.push_state_update()
+        except DeviceTimeoutError:
+            _LOGGER.warning(
+                "Timeout send state update to: %s (%s)",
+                self.name,
+                self._device.device_info,
+            )
+        except Exception:  # pylint: disable=broad-except
+            _LOGGER.exception(
+                "Unknown exception caught while sending state update to: %s (%s)",
+                self.name,
+                self._device.device_info,
+            )
+
+    @property
+    def device_info(self):
```

Comment:
Are any of the members here and below actually used? If not, i would suggest removing them for now given they seem to mirror what the entities already do by asking the device itself for properties rather than reading them from the coordinator.

Reply:
I have a preference for the wrapper pattern around external libraries, this helps prevent unintentionally abstraction bleeds and promotes DRY principals and the introduction of the coordinator enables centralized of device library code to this namespace.

That being said, I will look at the impact of removing just these properties, since they are likely to be uniquely used across platform classes.

Reply:
I am supportive of wrapper library following the HA best practices docs.  What I mean is difference between coordinator and entity.  I don't see many of these methods used by entities or anywhere unless I'm missing it.

Reply:
These properties are actually what climate.py and switch.py is calling.

```python
class GreeSwitchEntity(CoordinatorEntity, SwitchEntity):
    """Representation of a Gree HVAC device."""

    def __init__(self, device):
        """Initialize the Gree device."""
        super().__init__(device)
        self._device = device
        self._name = device.device_info.name + " Panel Light"
        self._mac = device.device_info.mac
        self._icon = "mdi:lightbulb"
```

**device** is the coordinator in this context. So this property on **SwitchEntity** is accessing the property on the coordinator, which wraps the device library implementation:

```python
    @property
    def is_on(self) -> bool:
        """Return if the light is turned on."""
        return self._device.light
```

Reply:
I see! Thank you for walking me through that.  I definitely missed that given the variable naming.

Given that all of these methods are 1:1 with the device, I would recommend using the coordinator to refresh the device only, but not wrap all of the methods That is, both the coordinator and the entity can keep a reference to the device, but the coordinator calls "self._device.update_state()".  The entity can just access _device the same way it does now.  

In order to do that, you'd need to hold on to the devices to pass both the device and the coordinator to the entity, but in exchance >120 lines of code here can be removed.  I think also differentiating "device" from "coordinator" can also be a readability improvement, e.g. given the mistake i mead when reading this.